### ISSUE TITLE
PeptideShaker MzIdentML export issue

### ISSUE THREAD

**Issue by vladie0** [compomics/peptide-shaker#118]
Dear developers, 

While exporting results (obtained from searchGUI + PeptideShaker) to MzIdentML I have encountered issues when reading these MzIdentML by third party software (openMS,pyteomics).

In order to verify whether the problem is caused by PeptideShaker or my search parameters, I have exported the PeptideShaker example file to MzIdentML format and attempted to read in the MzIdentML files. 

Checking the semantic integrity of the converted PeptideShaker example MzIdentML file with OpenMS integrity checking utility (http://ftp.mi.fu-berlin.de/pub/OpenMS/release1.10-documentation/html/classOpenMS_1_1MzIdentMLFile.html) I have encountered the following error : 

Error: whitespace expected( in line 148616 column 115)

Inspecting this line in the MzIdentML file, I have pinpointed the problem to this line:

```
<cvParam cvRef="PSI-MS" accession="MS:1000796" name="spectrum title" value="qExactive01819.6511.6511.2 File:"qExactive01819.raw", NativeID:"controllerType=0 controllerNumber=1 scan=6511""/>
```

Deleting this line from the MzIdentML shifts the error to the next cvParams line.

Furthermore, exporting my own PeptideShaker analysis results to MzIdentML and importing with openMS results in the following import error:

```
the element 'modification index (M)' could not be found
<Misplaced elements ignored in 'ParamGroup' in Fragmentation> occurred 64807 times
```

I'm happy to provide you with additional information/ help solve this problem.
The alternative for me would be to write my own MzIdentML parser for PeptideShaker files which would be quite exhaustive.
Thanks for you help

**Comment by hbarsnes** [comment]
Thanks for telling us about this. It seems like we no longer correctly escape the quotes in the spectrum titles. We used to do this before, so it must have been lost as part of another update. I'll look into it and get back to you. Hopefully that will solve the first issue.

As for "the element 'modification index (M)' could not be found", this could refer to a new mzIdentML 1.2 feature that should not be included when exporting mzIdentML 1.1. So I'll remove it from the default export. Perhaps you could try to manually delete the modification index cv terms in the SearchModification elements inside ModificationParams to see if that gets rid of the error? Or if this is actually referring to something else?

And is the "Misplaced elements ignored in 'ParamGroup' in Fragmentation" a separate error or is it linked to the modification index issue? From your message it is not clear which element that is misplaced? Maybe some of the IonType cv terms used are not supported in OpenMS?

**Comment by hbarsnes** [comment]
Here's an updated mzIdentML file of the tutorial dataset with the modification index cv terms removed and the quotes in the spectrum titles correctly escaped. Can you check if you're able to open this one without any issues?

[tutorial.zip](https://github.com/compomics/peptide-shaker/files/97336/tutorial.zip)

**Comment by vladie0** [comment]
First of all, Thank you for you support !

The tutorial MzIdentML you provided passed the OpenMS integrity test. 
However while processing the MzIdentlML file I get the following error:

RuntimeError: the element 'Glu->pyro-Glu (E)' could not be found

I suspect that OpenMS itself is responsible for this issue, and will inspect this. I will update you when I find the source of this issue.

Thank you again for your outstanding support

**Comment by vladie0** [comment]
Would it be possible to provide me with an updated version of PeptideShaker, so I could test this on my own results ? This will allow me to determine the source of the issue.

**Comment by hbarsnes** [comment]
> Would it be possible to provide me with an updated version of PeptideShaker, so I could test this on my own results ? This will allow me to determine the source of the issue.

Sure, here is a beta version with the above mzIdentML fixes: https://www.dropbox.com/s/lw3rzr6mgjkjes5/PeptideShaker-1.4.1-beta.zip?dl=0

It also contains some other minor changes, but these should not affect your testing.

> RuntimeError: the element 'Glu->pyro-Glu (E)' could not be found

I would guess that this is more html escaping issues? As in the file this one is listed as "Gln-&gt;pyro-Glu". This should be consistent in the exported mzIdentML file though.

Let me know what you find out.

**Comment by vladie0** [comment]
I reprocessed my data using the PeptideShaker beta version.
Both the PeptideShaker example file and my own data is processed correctly according to my own scripts ( I need MzIdentML files for post-processing). 

However the "RuntimeError: the element 'Glu->pyro-Glu (E)' could not be found" persists when attempting to read the files with openMS

I have narrowed the problem down to the following line in the MzIdentML files : 

"<Modification monoisotopicMassDelta="-17.026549" residues="Q" location="0" >
				<cvParam cvRef="UNIMOD" accession="UNIMOD:28" name="Gln-&gt;pyro-Glu" value="-17.026549"/>
</Modification>"

However, I have only used openMS as a verification tool to test whether there was a problem with the MzIdentML files, I do not need it for my own purposes.

Thank you for your excellent support

**Comment by hbarsnes** [comment]
Great! I guess we can just ignore the Open MS issue for now then. But good to know in case someone needs to use OpenMS in the future.

I will leave the issue open until the new PeptideShaker version has been released. It should be available by the end of the week.

**Comment by hbarsnes** [comment]
The issue has been fixed with the release of PeptideShaker 1.5.0. Please update to this version instead of the beta release as it also contains some other important fixes and improvements.
